### PR TITLE
tsweb: extra logging fields for HTTP handlers

### PR DESCRIPTION
**Pull request by crawshaw**
Signed-off-by: David Crawshaw <crawshaw@tailscale.com>

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/548)

### CODE REVIEW

File: tsweb/log.go [hunk 1]
```diff
@@ -46,6 +46,12 @@ type AccessLogRecord struct {
 	Bytes int `json:"bytes"`
 	// Error encountered during request processing.
 	Err string `json:"err"`
+
+	// Extra contains extra fields logged by the HTTP handler.
+	//
+	// It is called "x" in the JSON both for brevity and so
+	// it comes last.
+	Extra map[string]string `json:"x,omitempty"`
```
Comment (danderson) [inline tsweb/log.go:54]: Why? It feels weird to add a grab bag of "stuff" to logs without knowing what to expect.

Also not a fan of "x" as a field name, to me that looks mysterious if you're looking at logs.

I'm not opposed enough to block any of this, it's more of a "feels weird" than principled objection. What are you thinking of using this for?

Comment (crawshaw) [reply tsweb/log.go:54]: My objective is to let handlers add handler-specific information about a HTTP request to the log. For example, after decoding, a particular query parameter may contain the important information about this request is doing.

There's a tension between teaching the log type everything about all handlers, and capturing generic fields. I've seen both and can speak to the tradeoffs. There's no obviously right solution.

One thing I could do is the compromise we ended up with at Google:

```Extra interface{}```

with each handler (or better, server) defining a type detailing all of its logging fields. Would you prefer that?

Comment (bradfitz) [reply tsweb/log.go:54]: I guess where we don't want to end up is a bunch of untyped strings everywhere that can't be found easily and have them stepping on each other's toes fighting in the same cluttered namespace.

I think either enforcing a registry of known extras keys and/or enforcing a max-one-extra-per-Handler with types like your `Extra interface{}` both help.

Or just a `type ExtraKey string` and making it `Extra map[ExtraKey]interface{}`

I'm not particularly concerned which route we take.

**Comment by bradfitz** [comment]
Closing for now until needed again.

**Pull request closed without merging**